[Bug]: Cards not compatible with React 18

### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
System:
    OS: Windows 10 10.0.22621
    CPU: (8) x64 Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz
    Memory: 16.60 GB / 31.59 GB
  Browsers:
    Edge: Spartan (44.22621.1105.0), Chromium (109.0.1518.70)
```


### Are you reporting Accessibility issue?

no

### Reproduction

https://codesandbox.io/s/async-wood-bd02mn?file=/src/App.tsx

### Bug Description

## Actual Behavior
Page renders blank

## Expected Behavior
Render the UI

### Logs

```shell
Warning: React has detected a change in the order of Hooks called by CardHeader. This will lead to bugs and errors if not fixed. For more information, read the Rules of Hooks: https://reactjs.org/link/rules-of-hooks

   Previous render            Next render
   ------------------------------------------------------
1. useContext                 useContext
2. useRef                     useRef
3. useContext                 useContext
4. useContext                 useContext
5. useId                      useEffect
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    at eval (https://bd02mn.csb.app/node_modules/@fluentui/react-card/lib-commonjs/components/CardHeader/CardHeader.js:15:33)
    at div
    at eval (https://bd02mn.csb.app/node_modules/@fluentui/react-card/lib-commonjs/components/CardPreview/CardPreview.js:15:34)
    at div
    at eval (https://bd02mn.csb.app/node_modules/@fluentui/react-card/lib-commonjs/components/Card/Card.js:16:27)
    at App (https://bd02mn.csb.app/src/App.tsx:29:29)
```


### Requested priority

Blocking

### Products/sites affected

_No response_

### Are you willing to submit a PR to fix?

yes

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

@sopranopillow  QQ did you happen to fix the issue with the dependencies that seems to be related?
I wasn't able to capture the below issue in the repro space since it doesn't expose the NPM install command.

```
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN Found: @types/react@18.0.27
npm WARN node_modules/@types/react
npm WARN   dev @types/react@"^18.0.27" from the root project
npm WARN   55 more (@fluentui/foundation-legacy, @fluentui/react, ...)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer @types/react@">=16.8.0 <18.0.0" from @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN node_modules/@fluentui/react-components/node_modules/@fluentui/react-virtualizer
npm WARN   @fluentui/react-virtualizer@"9.0.0-alpha.3" from @fluentui/react-components@9.14.0
npm WARN   node_modules/@fluentui/react-components
npm WARN 
npm WARN Conflicting peer dependency: @types/react@17.0.53
npm WARN node_modules/@types/react
npm WARN   peer @types/react@">=16.8.0 <18.0.0" from @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN   node_modules/@fluentui/react-components/node_modules/@fluentui/react-virtualizer
npm WARN     @fluentui/react-virtualizer@"9.0.0-alpha.3" from @fluentui/react-components@9.14.0
npm WARN     node_modules/@fluentui/react-components
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN Found: @types/react-dom@18.0.10
npm WARN node_modules/@types/react-dom
npm WARN   dev @types/react-dom@"^18.0.10" from the root project
npm WARN   43 more (@fluentui/react, @fluentui/react-accordion, ...)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer @types/react-dom@">=16.8.0 <18.0.0" from @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN node_modules/@fluentui/react-components/node_modules/@fluentui/react-virtualizer
npm WARN   @fluentui/react-virtualizer@"9.0.0-alpha.3" from @fluentui/react-components@9.14.0
npm WARN   node_modules/@fluentui/react-components
npm WARN 
npm WARN Conflicting peer dependency: @types/react-dom@17.0.18
npm WARN node_modules/@types/react-dom
npm WARN   peer @types/react-dom@">=16.8.0 <18.0.0" from @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN   node_modules/@fluentui/react-components/node_modules/@fluentui/react-virtualizer
npm WARN     @fluentui/react-virtualizer@"9.0.0-alpha.3" from @fluentui/react-components@9.14.0
npm WARN     node_modules/@fluentui/react-components
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN Found: react@18.2.0
npm WARN node_modules/react
npm WARN   react@"^18.2.0" from the root project
npm WARN   60 more (@fluentui/foundation-legacy, @fluentui/react, ...)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer react@">=16.8.0 <18.0.0" from @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN node_modules/@fluentui/react-components/node_modules/@fluentui/react-virtualizer
npm WARN   @fluentui/react-virtualizer@"9.0.0-alpha.3" from @fluentui/react-components@9.14.0
npm WARN   node_modules/@fluentui/react-components
npm WARN
npm WARN Conflicting peer dependency: react@17.0.2
npm WARN node_modules/react
npm WARN   peer react@">=16.8.0 <18.0.0" from @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN   node_modules/@fluentui/react-components/node_modules/@fluentui/react-virtualizer
npm WARN     @fluentui/react-virtualizer@"9.0.0-alpha.3" from @fluentui/react-components@9.14.0
npm WARN     node_modules/@fluentui/react-components
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN Found: react-dom@18.2.0
npm WARN node_modules/react-dom
npm WARN   react-dom@"^18.2.0" from the root project
npm WARN   44 more (@fluentui/react, @fluentui/react-accordion, ...)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer react-dom@">=16.8.0 <18.0.0" from @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN node_modules/@fluentui/react-components/node_modules/@fluentui/react-virtualizer
npm WARN   @fluentui/react-virtualizer@"9.0.0-alpha.3" from @fluentui/react-components@9.14.0
npm WARN   node_modules/@fluentui/react-components
npm WARN
npm WARN Conflicting peer dependency: react-dom@17.0.2
npm WARN node_modules/react-dom
npm WARN   peer react-dom@">=16.8.0 <18.0.0" from @fluentui/react-virtualizer@9.0.0-alpha.3
npm WARN   node_modules/@fluentui/react-components/node_modules/@fluentui/react-virtualizer
npm WARN     @fluentui/react-virtualizer@"9.0.0-alpha.3" from @fluentui/react-components@9.14.0
npm WARN     node_modules/@fluentui/react-components
```

--- a/change/@fluentui-react-provider-2dd88fac-fd7c-4313-98b8-3941d9399f42.json
+++ b/change/@fluentui-react-provider-2dd88fac-fd7c-4313-98b8-3941d9399f42.json
@@ -1,0 +1,7 @@
+{
+  "type": "patch",
+  "comment": "fix: Do not add an escape character for colons in the styleTagId since useId now removes colons from the generated id in React 18.",
+  "packageName": "@fluentui/react-provider",
+  "email": "esteban.230@hotmail.com",
+  "dependentChangeType": "patch"
+}

--- a/change/@fluentui-react-utilities-db7dae74-67ca-40d9-9548-f9b0965366c2.json
+++ b/change/@fluentui-react-utilities-db7dae74-67ca-40d9-9548-f9b0965366c2.json
@@ -1,0 +1,7 @@
+{
+  "type": "patch",
+  "comment": "fix: Generate id first to avoid hook order mismatch in React 18 and remove colons from generated id in React 18.",
+  "packageName": "@fluentui/react-utilities",
+  "email": "esteban.230@hotmail.com",
+  "dependentChangeType": "patch"
+}

--- a/packages/react-components/react-provider/src/components/FluentProvider/useFluentProviderThemeStyleTag.ts
+++ b/packages/react-components/react-provider/src/components/FluentProvider/useFluentProviderThemeStyleTag.ts
@@ -62,8 +62,7 @@ export const useFluentProviderThemeStyleTag = (options: Pick<FluentProviderState
       : '';
   }, [theme]);
 
-  // When using React 18, the id generated will contain : which is not valid unless we add an escape character
-  const rule = `.${styleTagId.replace(/:/g, '\\:')} { ${cssVarsAsString} }`;
+  const rule = `.${styleTagId} { ${cssVarsAsString} }`;
 
   useInsertionEffect(() => {
     styleTag.current = createStyleTag(targetDocument, { ...styleElementAttributes, id: styleTagId });

--- a/packages/react-components/react-utilities/src/hooks/useId.ts
+++ b/packages/react-components/react-utilities/src/hooks/useId.ts
@@ -37,7 +37,9 @@ export function useId(prefix: string = 'fui-', providedId?: string): string {
   const _useId: () => string | undefined = (React as never)['use' + 'Id'];
 
   if (_useId) {
-    return providedId || `${idPrefix}${prefix}${_useId()}`;
+    const generatedId = _useId()?.replace(/:/g, '');
+
+    return providedId || `${idPrefix}${prefix}${generatedId}`;
   }
 
   // Hooks appear to be running conditionally, but they will always run in the same order since it's based on

--- a/packages/react-components/react-utilities/src/hooks/useId.ts
+++ b/packages/react-components/react-utilities/src/hooks/useId.ts
@@ -34,12 +34,15 @@ export function useId(prefix: string = 'fui-', providedId?: string): string {
 
   // Checking if useId is available on React, if it is, we use it to generate the id. String concatenation is used to
   // prevent bundlers from complaining with older versions of React.
-  const _useId: () => string | undefined = (React as never)['use' + 'Id'];
+  const _useId = (React as never)['use' + 'Id'] as (() => string) | undefined;
 
   if (_useId) {
-    const generatedId = _useId()?.replace(/:/g, '');
+    const generatedId = _useId();
 
-    return providedId || `${idPrefix}${prefix}${generatedId}`;
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    const escapedId = React.useMemo(() => generatedId.replace(/:/g, ''), [generatedId]);
+
+    return providedId || `${idPrefix}${prefix}${escapedId}`;
   }
 
   // Hooks appear to be running conditionally, but they will always run in the same order since it's based on

